File: Problems/LongestSubstringWithAtMostKDistinctCharacters/LongestSubstringWithAtMostKDistinctCharacters.cpp
class Solution {
public:
    int lengthOfLongestSubstringKDistinct(string s, int k) {
        int windowStart = 0;
        int maxSize = 0;
        unordered_map<char, int> charFrequencyMap;
        
        for(int windowEnd = 0; windowEnd < s.length(); windowEnd++) {
            char rightChar = s[windowEnd];
            charFrequencyMap[rightChar]++;
            
            while((int)charFrequencyMap.size() > k) {
                char leftChar = s[windowStart];
                charFrequencyMap[leftChar]--;
                if(charFrequencyMap[leftChar] == 0) {
                    charFrequencyMap.erase(leftChar);
                }
                windowStart++;
            }
            maxSize = max(maxSize, windowEnd - windowStart + 1);
            
                
        }
        return maxSize;
            
        
    }
};

File: Problems/LongestSubstringWithAtMostKDistinctCharacters/LongestSubstringWithAtMostKDistinctCharacters.py
'''
Grow window until > k distinct characters. Then shrink to satisfy
'''

class Solution:
    def lengthOfLongestSubstringKDistinct(self, s: str, k: int) -> int:
        maxSubarray = 0
        start = 0
        d = {}
        
        for end, char in enumerate(s):
            d[char] = d.get(char, 0) + 1
            
            while len(d) > k:
                leftChar = s[start]
                
                d[leftChar] -= 1
                if d[leftChar] == 0:
                    del d[leftChar]
                    
                start += 1
            
            maxSubarray = max(maxSubarray, end - start + 1)
        
        return maxSubarray
            
            
        
        
